[[ VarAccess ]] Cleanup and optimization

runrevmark opened the pull request:
This is work-in-progress set of st-git patches which cleans up the implementation of variable access in the engine.

At present it includes the following:

- Basic (incomplete) tests and benchmarks
- Removal of some unused virtual methods from MCExpression (and descendants)
- Changes uses of MCContainer to be mostly allocated 'on stack'
- Moves all code to use just 'evalcontainer' which removes the 'evalvar' virtual method.
- Unifies various hashing code from different places
- Refactors code in MCName to do lookup and creation into a template based approach
- Adds optimized code paths for creating 'index' names (integers which fit into index_t)
- Ensures index name code paths are used when looking up array indices.
- Adds the idea of purity to MCExpression (a pure MCExpression is one which has no side effects)
- Uses templates to allow more efficient code paths for common MCVarref operations
- Implements 'pure' code paths for set(into), give_value(into), dofree() and eval_ctxt MCVarref operations

This patch provides a *substantial* speed up to variable manipulation operations. Indeed, indexing operations using strings or literals (names) are 2-3 times faster (both storing and fetching); whilst indexing operations using integers are 4 times faster (storing) and 7 times faster (fetching). (All timings compared against 9dp4).

Note: This additionally changes the external interfaces so that internally they act on a container, rather than a variable - this should mean that externals should now be able to correctly set variables which are references to array elements... Tests need to be added for the externals interfaces to ensure that nothing has broken!

Note: The 'before' and 'after' operations still need optimized codepaths to be implemented; this should be combined with a refactoring of the code which manipulates Data and String values in this way since it is currently spread out in several places and the semantics are not 100% clear.

Note: There remains a fault with case-sensitivity being remembered for reference parameters - they are currently evaluated with the caseSensitive of the callee, rather then when the path was evaluated in the caller.

Note: The 'pure' dofree() code path is currently broken (I thought it might be because it is creating nested arrays when not necessary, but changing that does not seem to fix things - the IDE won't run properly with that code path enabled).

Note: For some reason one of the matchText tests fails without ensuring 'clearuql()' is done in MCVariable::setvalueref. There must be a subtle change somewhere which means the 'v0' undeclared var in the test is *not* getting cleared whereas it did before.

runrevmark committed:
[[ VarAccess ]] Remove 'rootmatches()' and uses from MCVarref

This patch removes the 'rootmatches()' method from MCVarref along
with its uses in the arithmetic commands. The latter used it to
compute overlap between base and value expressions but this
information is no longer needed.

runrevmark committed:
[[ VarAccess ]] Make MCContainer creation act on a reference

This patch changes the creation of MCContainers so that the
constructors act on a reference, rather than returning a new
object. This allows stack allocation of MCContainer classes in
almost all cases.

runrevmark committed:
[[ VarAccess ]] Remove 'evalvar()' method

This patch removes the 'evalvar()' virtual method from MCExpression
replacing all uses with 'evalcontainer()'.

runrevmark committed:
[[ VarAccess ]] Add tests and benchmarks

This patch adds some tests and some benchmarks covering
variable access operations.

peter-b commented on engine/src/cmdse.cpp:
```diff
@@ -623,8 +622,8 @@ void MCDispatchCmd::exec_ctxt(MCExecContext &ctxt)
-        MCContainer *t_container;
-        if (tptr -> evalcontainer(ctxt, t_container))
+        MCContainer *t_container = new(nothrow) MCContainer;
+        if (tptr -> evalcontainer(ctxt, *t_container))
```
At the moment, this will leak the container if `evalcontainer()` fails.  I suggest an `MCAutoPointer<MCContainer>` here.

peter-b commented on engine/src/exec-keywords.cpp:
```diff
@@ -149,8 +149,8 @@ void MCKeywordsExecCommandOrFunction(MCExecContext& ctxt, bool resolved, MCHandl
-        MCContainer *t_container;
-        if (tptr -> evalcontainer(ctxt, t_container))
+        MCContainer *t_container = new(nothrow) MCContainer;
+        if (tptr -> evalcontainer(ctxt, *t_container))
```
This will also leak the allocated `MCContainer` instance.

peter-b commented on engine/src/externalv1.cpp:
```diff
@@ -967,44 +967,17 @@ Exec_stat MCExternalV1::Handle(MCObject *p_context, Handler_type p_type, uint32_
-		}
+    {
+        MCContainer *t_container;
+        t_container = p_parameters -> eval_argument_container();
```
Please use a managed lifetime here if possible!

runrevmark replied:
I'm not sure it is possible as the MCParameter owns the MCContainer, and the MCReferenceExternalVariable will share it temporarily.

peter-b commented on engine/src/externalv0.cpp:
```diff
@@ -1441,29 +1454,38 @@ static char *get_array_utf8(const char *arg1, const char *arg2,
+        return NULL;
+    }
+    
+    MCArrayRef t_array = (MCArrayRef)*t_value;
```
This means that there are two pointers to the value live on the stack at the same time, one managed and one unmanaged. I suggest something like:

```C++
MCAutoArrayRef t_array = static_cast<MCArrayRef>(t_value.Take());
```

Using `Take()` avoids frobbing the reference count.

Also, don't forget that we're supposed to be avoiding C-style casts these days! ;-) There are a couple of examples like this.

runrevmark replied:
Hmmm - I'm not sure this works with the current definition of the MCArrayRef constructor in MCAutoValueRefBase as it retains the incoming argument. Here we don't want to fettle with reference counts at all.

peter-b replied:
Good point. I suppose the alternative would be:

```
MCAutoArrayRef t_array = static_cast<MCArrayRef>(*t_value);
```

Admittedly this does require taking another reference to the value, but that's not necessarily a bad thing.

runrevmark replied:
Well, it is a bad thing as the operation is atomic so the reference count bump is entirely unnecessary. I'd rather we 'fixed' the auto classes to allow movement of a MCValueRef into another auto container or a specific MC<type>Ref container - I'm just not sure what the 'correct' pattern for that is.

peter-b replied:
> I'd rather we 'fixed' the auto classes to allow movement of a MCValueRef into another auto container or a specific MCRef container

Good idea.  I'll think of something. 👍

peter-b commented on engine/src/variable.cpp:
```diff
@@ -1114,6 +1114,24 @@ MCVarref *MCVariable::newvarref(void)
+        r_out[i + p_base.size()] = p_extra[i];
+    
+    return true;
+}
```
Consider:

```c++
static MCAutoPointer<MCNameRef[]>
__join_paths(MCSpan<MCNameRef> p_base,
             MCSpan<MCNameRef> p_extra)
{
    MCAutoPointer t_result =
        new (nothrow) MCNameRef[p_base.size() + p_extra.size()];
    if (t_result)
    {
        /* Populate t_result */
    }
    return std::move(t_result);
}
```

Then, when you use it, you can write:

```c++
MCAutoPointer<MCNameRef[]> t_full_path =
    __join_paths(MCMakeSpan(m_path, m_length),
                 MCMakeSpan(p_path, p_path_length));
if (!t_full_path)
    return false;
```

This ensures RAII everywhere, and it's the pattern recommended by the C++ Guideline Standard for functions that create an owned value.

BTW I generally recommend `MCMakeSpan()` for creating spans in argument contexts like this.

peter-b reviewed (commented):
This is awesome. 🥇

While you're at it, why not convert `MCContainer::m_path` to an `MCAutoNameRefArray` or something?

runrevmark committed:
[[ Foundation ]] Unify hash function implementations

This patch moves all implementations of hash functions to inline
implementations in a private header foundation-hash.h.

Hash implementations are provided for integer-like things,
byte-like things and Unicode codepoint type things.

peter-b commented:
@runrevmark Did you intentionally include the hash functions refactor in this PR?

runrevmark commented:
@peter-b: Yes - it is an st-git stack so I can separate things out subsequently if necessary. The hash cleanup is required for an optimization to looking up elements in sequences.

runrevmark committed:
[[ Foundation ]] Renamed MCNameLookup method

The MCNameLookup method has been renamed to MCNameLookupCaseless
to more accurately describe its function.

runrevmark committed:
[[ Foundation ]] Add MCStringIsEqualToChars

This patch adds a MCStringIsEqualToChars function to libfoundation
allowing direct comparison between a MCStringRef and buffer of
unichar_t.

runrevmark committed:
[[ Foundation ]] Improve notion of index name

This patch improves the performance of names when used as
indexes into arrays. It adds an optimized codepath for looking
up and creating names which are integers (in the range supported
by index_t) and ensures these codepaths are used when converting
from a value to a nameref if the initial value is a number.

runrevmark committed:
[[ Expression ]] Add notion of purity to MCExpression

This patch adds a virtual method 'is_pure' to the MCExpression
class. This method returns true if evaluating the expression
has no side-effects.

Note: This is currently incomplete - only MCLiteral, MCConstant
and MCVarref have a non-default implementation.

runrevmark committed:
[[ Foundation ]] Add MCArrayMutateValue function

This patch adds an MCArrayMutateValue function to libfoundation. The
function allows direct access to the slot of an element within an
array, allowing the caller to assign to it directly, rather than
having to fetch and then store.

runrevmark committed:
[[ Foundation ]] Add capacity field to MCAutoArrayRefBase

This patch adds a 'capacity' field to MCAutoArrayRefBase allowing
the auto class to either be used as a fixed size array, or a
dynamically sizable array more easily.

runrevmark committed:
[[ VarAccess ]] Improve access operations

This patch rewrites the resolve method to use a template class which
is customizable.

Customizations have been implemented for:

- deferring evaluation of array keys (i.e. storing the full list of
  MCNameRefs)

- fetching an array element when the dimensions are all pure (eval_ctxt)

- storing into an array element when the dimensions are all pure (set and
  give_value)

- removing an array element when the dimensions are all pure (dofree)

Note: 'before' and 'after' mutation operations do not currently have a
'pure' code path.

runrevmark committed:
[[ WIP ]] Temporary fixes

This removes the pure path for dofree() and ensures a var
being set via setvalueref is clearuql'd first (otherwise
the matchText tests fail).

peter-b commented on engine/src/variable.cpp:
```diff
@@ -1114,6 +1114,24 @@ MCVarref *MCVariable::newvarref(void)
+            t_result[i + p_base.size()] = p_extra[i];
+    }
+    
+    return std::move(t_result);
```
I gave you duff information here. `return` implicitly moves its argument, so this can just be:

```c++
return t_result;
```

peter-b commented on engine/src/variable.cpp:
```diff
@@ -1186,27 +1186,25 @@ bool MCContainer::set_real(double p_real)
+	r_container.m_variable = p_var;
+	r_container.m_length = 0;
+    r_container.m_path = nil;
+	r_container.m_case_sensitive = false;
```
This could also:

```c++
createwithpath(p_var, nullptr, 0, r_container;
```
